### PR TITLE
fix(issue): [Bug] Tier resolver ignores session model: standard units get cheapest, light units hop to claude-haiku

### LINKED ISSUE #810: [Bug] Tier resolver ignores session model: standard units get cheapest, light units hop to claude-haiku
## Problem

When a user runs auto-mode on a **non-Claude provider** (e.g. a GLM model set via `/model`), GSD does not actually use the user's selected model for many units. Two related defects in `model-router.js` cause silent model switches:

1. **Standard/heavy units get the cheapest model, not the selected one.** `execute-task` / `plan-slice` / `research` resolve to the `standard` tier. All GLM models are classified `standard` (via the D-15 unknown-model default), `getEligibleModels` sorts them **cheapest-first**, and `findModelForTier` returns `eligible[0]` — so `glm-4.5-air` (first/cheapest in the ZAI provider list) is picked instead of the user's selected `glm-5.2`. The session/selected model is never consulted.

2. **Light-tier units hop to the Anthropic canonical model.** `complete-slice` (closeout) resolves to the `light` tier. The ZAI provider has no `light`-classified model, so `resolveModelForTier` falls back to the hardcoded canonical `claude-haiku-4-5`. Because the user is not authenticated to Claude, the unit fails with `"Not logged in · Please run /login"`, produces **ghost completions** (0 tool calls, 3–4s), and the state machine re-dispatches in a stuck loop (5× observed for `complete-slice/M017/S01`).

Both defects share a root design gap: **the tier resolver has no concept of the user's selected/session model.** It picks by tier + price (or canonical fallback), and only preserves the session model by luck when the selected model happens to be the cheapest in its tier.

### Defect 1 — cheapest-first ignores session model

`getEligibleModels` (`model-router.js:229-249`) auto-detect branch:
```js
// 2. Auto-detect: filter by tier, sort cheapest first
return availableModelIds
    .filter(id => getModelTier(id) === tier)
    .sort((a, b) => {
        const costA = getModelCost(a);
        const costB = getModelCost(b);
        return costA - costB;          // ← cheapest first, no session-model preference
    });
```

`findModelForTier` (`model-router.js:459-471`) takes the first element:
```js
function findModelForTier(tier, routingConfig, availableModelIds, crossProvider) {
    const eligible = getEligibleModels(tier, availableModelIds, routingConfig);
    if (eligible.length === 0) return undefined;
    if (crossProvider) {
        return eligible[0];            // ← cheapest, not the user's selected model
    }
    const sameProvider = eligible.filter(id => { /* claude-* canonical */ });
    return sameProvider[0];            // ← same issue on the same-provider path
}
```
There is no parameter or signal for "the model the user picked via `/model`". On a provider where multiple models share a tier (e.g. ZAI's `glm-4.5-air`, `glm-5.1`, `glm-5.2` all = `standard`), the cheapest wins regardless of user intent.

### Defect 2 — canonical fallback cross-hops to Anthropic

`complete-slice` → phase chain `["completion"]` (`unit-registry.js:170-173`) → tier `light` for **all** profiles (`preferences-models.js` `PROFILE_TIER_MAP`: `completion: "light"` in budget/balanced/quality). `DEFAULT_TOKEN_PROFILE = "balanced"` applies implicitly even with no `token_profile` configured, so profile defaults are injected.

`resolveProfileDefaults` builds `models.completion = resolveModelForTier("light", <scoped models>, routingConfig)`. `resolveModelForTier` (`model-router.js:474-493`):
```js
const resolved = findModelForTier(tier, routingConfig, availableModelIds, allowCrossProvider);
if (resolved) {
    return normalizeResolvedTierModelId(resolved, tier, routingConfig);
}
incrementLegacyTelemetry("legacy.providerDefaultUsed");
return canonicalModelForTier(tier);   // ← BUG: cross-provider hop to claude-haiku-4-5
```
`CANONICAL_TIER_MODELS` (`model-router.js:441-444`):
```js
const CANONICAL_TIER_MODELS = {
    light: "claude-haiku-4-5",        // ← hardcoded Anthropic fallback
    standard: "claude-sonnet-4-6",
    heavy: "claude-opus-4-6",
};
```
When the anchor provider has no `light`-classified model, the final fallback unconditionally returns the Anthropic model, ignoring both the anchor-provider scoping and the user's session model. This contradicts the documented design in `preferences-models.js`: *"token profiles do not hop to a cheaper provider."*

The profile-derived config is then stamped `source: "explicit"` (`auto-model-selection.js:329-335`), which **disables dynamic routing** (`auto-model-selection.js:465`), so routing cannot rescue the dispatch.

### Relationship to existing reports

This issue supersedes and generalizes #757. #757 covers only the empty-`availableModelIds` early return at `model-router.js:482-484`. This issue covers (a) the non-empty-list no-tier-match branch at `model-router.js:493` (defect 2) and (b) the cheapest-first selection at `model-router.js:242-248` + `463`/`471` (defect 1).

## Expected Behavior

The tier resolver should prefer the user's **selected/session model** whenever it satisfies the tier requirement, and must never silently hop to a hardcoded Anthropic model on a non-Claude provider.

Concrete fix suggestions:

**For defect 1** — thread the session model into tier resolution and prefer it when tier-compatible:
```diff
 export function getEligibleModels(tier, availableModelIds, routingConfig, sessionModelId) {
     // ... explicit tier_models handling unchanged ...
+    const filtered = availableModelIds.filter(id => getModelTier(id) === tier);
+    // Prefer the user's selected model when it qualifies for the tier
+    if (sessionModelId && filtered.some(id => sameModel(id, sessionModelId))) {
+        return [sessionModelId, ...filtered.filter(id => !sameModel(id, sessionModelId))];
+    }
     return filtered.sort((a, b) => getModelCost(a) - getModelCost(b));
 }
```
`findModelForTier` / `resolveModelForTier` would pass the session/anchor model id through, and callers in `preferences-models.js` / `auto-model-selection.js` would supply it from `ctx.model` / `autoModeStartModel`.

**For defect 2** — do not cross-hop when same-provider resolution was requested (`model-router.js:491-493`):
```diff
   const resolved = findModelForTier(tier, routingConfig, availableModelIds, allowCrossProvider);
   if (resolved) {
       return normalizeResolvedTierModelId(resolved, tier, routingConfig);
   }
+  // Canonical fallback is itself a cross-provider hop to Anthropic. Only apply
+  // it when cross-provider resolution is permitted; otherwise yield undefined
+  // so the caller falls back to the user's session/anchor model.
+  if (!allowCrossProvider) {
+      return undefined;
+  }
   incrementLegacyTelemetry("legacy.providerDefaultUsed");
   return canonicalModelForTier(tier);
```
Plus the profile-default resolution path should pass `cross_provider: false` so the closer falls through to the session model instead of hopping to Claude. As a tier-upgrade alternative: when no model exists at the requested tier on the anchor provider, fall **up** (light → standard → heavy) on that provider rather than crossing providers.

## Environment
- GSD version: 1.3.0
- Provider: non-Claude (ZAI), selected model: `glm-5.2` (confirmed via metrics — 35 units on glm-5.2)
- Failing/affected units: `complete-slice` (defect 2, stuck loop); `execute-task` / `plan-slice` / `research` (defect 1, wrong-cheapest model)
- token_profile: none explicitly set → implicit `balanced` (D046)

## Reproduction Context
Auto-mode milestone closeout on a non-Claude provider. After all tasks in a slice complete, the `complete-slice` unit is dispatched, runs on `claude-haiku-4-5`, fails auth, and re-dispatches in a loop. Separately, standard-tier units silently run on the cheapest GLM variant rather than the selected one. Forensic metrics for the affected project:

| Unit type | Expected model | Actual model | Duration | Tool calls |
|-----------|----------------|--------------|----------|------------|
| execute-task / plan-slice / research | glm-5.2 | glm-4.5-air (cheapest) or glm-5.1 | 3–11 min | normal |
| complete-slice | glm-5.2 | claude-haiku-4-5 (canonical fallback) | 3–4 s | 0 (ghost) |

## Forensic Evidence
- Model distribution in `.gsd/metrics.json`: 174× `glm-5.1`, 135× `glm-5-turbo`, 35× `glm-5.2`, 7× `glm-4.5-air`, 5× `claude-haiku-4-5`, 3× `claude-sonnet-4-6`. The 5 `claude-haiku-4-5` entries exactly match the 5× stuck-loop dispatches of `complete-slice/M017/S01`.
- `complete-slice/M017/S01` activity logs (seq 383-385) all show `"Last reasoning: Not logged in · Please run /login"`, 0 tool calls, 3–4s duration (ghost completions).
- `execute-task`/`plan-slice`/`research` on the same slice ran to completion but on cheaper GLM variants, not the user's selected `glm-5.2`.
- `isSuspiciousGhostCompletion` exists in `auto-unit-closeout.js` (0 tool calls + short elapsed) — the ghost completions are detectable, but the state machine re-dispatches anyway.

## Immediate Workarounds (for affected users)
- **Covers both defects:** set `token_profile: burn-max` in PREFERENCES.md. `resolveProfileDefaults` returns no `models` block for burn-max and disables routing, so every unit falls through to the session model (`ctx.model`). `burn-max` does not skip phases.
- **For defect 2 only:** pin light phases explicitly in PREFERENCES.md (`models.completion`, `models.subagent`, `models.execution_simple` → the selected model id).

---
*Auto-generated by `/gsd forensics`*

### PR DESCRIPTION
**Pull request by jeremymcs**
## Summary
- Fixed session-model-aware tier/profile routing and verified syntax/whitespace; full tests were blocked by missing dependencies and offline npm.

## Bugs Addressed
- [x] **Standard-tier routing ignores the selected session model**
- [x] **Light-tier fallback silently switches to Anthropic**

## Verification
- Completed in the repository worktree before push.

## Related Issue
- Closes #810
- [#810 [Bug] Tier resolver ignores session model: standard units get cheapest, light units hop to claude-haiku](https://github.com/open-gsd/gsd-pi/issues/810)

## User
- @DragonSigh

## Repo
- `open-gsd/gsd-pi`

## Branch
- `issue/810-bug-tier-resolver-ignores-session-model--1782391321`

---

> [!NOTE]
> **Medium Risk**
> Changes core auto-mode model selection and profile default merging across many call sites; behavior shifts for multi-model providers and empty-registry edge cases, though coverage is added in model-router tests.
> 
> **Overview**
> Fixes **#810**: tier-based model selection now honors the user's **selected/session model** (`provider/id`) instead of always choosing the cheapest same-tier option or falling back to canonical Claude IDs when the registry has no tier match.
> 
> **Routing** — `getEligibleModels`, `resolveModelForComplexity`, and `resolveModelForTier` take an optional `preferredModelId`. When that model is available and meets the requested complexity tier, it is ranked first and can win over cheaper alternatives (e.g. GLM 5.2 over lighter GLM variants). `wasDowngraded` stays false when the preferred model matches the configured primary under different ID formatting.
> 
> **Preferences & bootstrap** — `preferredModelId` flows through `loadEffectiveGSDPreferences`, profile defaults, per-unit model resolution, auto `selectAndApplyModel`, orchestrator prefs, `resolveDefaultSessionModel` (with `sessionModelId`), and the config overlay so token profiles and implicit **balanced** defaults align with the session/auto-start snapshot.
> 
> **Fallback behavior** — With a non-empty provider list but no tier match, resolution prefers the session model or first available model rather than silently switching to Anthropic (notably for **light** tier).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 7cfce28af62abcea9590fd407e021ef6b3d7e986. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

---
<a href="https://app.blacksmith.sh/open-gsd/codesmith/gsd-pi/pr/854"><picture><source media="(prefers-color-scheme: dark)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-dark-v2.svg"><source media="(prefers-color-scheme: light)" srcset="https://pr-comments-assets.blacksmith.sh/codesmith/view-with-codesmith-light-v2.svg">View with Codesmith</picture></a>
<sup>Need help on this PR? Tag <code>/codesmith</code> with what you need. Autofix is enabled.</sup>

### CODE REVIEW

**Comment by jeremymcs** [comment]
✅ **Verification — all 2 tasks are addressed**

Re-checked PR #854 against [#810 [Bug] Tier resolver ignores session model: standard units get cheapest, light units hop to claude-haiku](https://github.com/open-gsd/gsd-pi/issues/810).

## Subtasks
- [x] **Standard-tier routing ignores the selected session model**
- [x] **Light-tier routing falls back to Anthropic on non-Claude providers**

_Re-run from Patchdeck to refresh._

File: src/resources/extensions/gsd/model-router.ts [hunk 1]
```diff
@@ -484,6 +493,19 @@ export function resolveModelForComplexity(
...
   }
 
+  const preferred = findPreferredModelForTier(requestedTier, availableModelIds, preferredModelId);
+  if (preferred && eligible.includes(preferred)) {
+    const fallbacks = buildFallbackChain(preferred, phaseConfig);
+    return {
+      modelId: preferred,
+      fallbacks,
+      tier: requestedTier,
+      wasDowngraded: preferred !== configuredPrimary,
```
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/gsd/model-router.ts]: **Accepted** — queued for a code change.
**In progress** — applying the accepted fix.
**Verifying** — checking the applied changes.
**Resolved** — addressed in commit `7cfce28`.

**Comment by jeremymcs** [comment]
Started an automated PR update.

Accepted review feedback is being applied.
Progress and final results will be posted in the related review thread(s).

**Comment by jeremymcs** [comment]
Started an automated PR update.

Accepted review feedback is being applied.
Progress and final results will be posted in the related review thread(s).

File: src/resources/extensions/gsd/preferences-models.ts [hunk 2]
```diff
@@ -477,26 +488,29 @@ const PROFILE_TIER_MAP: Record<TokenProfile, Record<string, ComplexityTier>> = {
...
+  preferredModelId?: string,
 ): Partial<GSDPreferences> {
   // burn-max never writes model defaults — preserve user-selected models.
   // For the other three profiles, derive concrete model IDs from the tier map
   // against the available-model list when the registry is provided. If callers
   // omit the registry entirely, use canonical fallbacks explicitly.
   const tierMap = PROFILE_TIER_MAP[profile];
   const resolveTierModel = (tier: ComplexityTier): string => Array.isArray(availableModelIds)
-    ? resolveModelForTier(tier, availableModelIds, routingConfig)
+    ? resolveModelForTier(tier, availableModelIds, routingConfig, undefined, preferredModelId)
```
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/gsd/preferences-models.ts]: **Accepted** — queued for a code change.
**In progress** — applying the accepted fix.
**Verifying** — checking the applied changes.
**Resolved** — addressed in commit `7cfce28`.

**Comment by jeremymcs** [comment]
Started an automated PR update.

Accepted review feedback is being applied.
Progress and final results will be posted in the related review thread(s).

File: src/resources/extensions/gsd/model-router.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/gsd/model-router.ts]: Addressed in commit `7cfce28`.

File: src/resources/extensions/gsd/preferences-models.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/gsd/preferences-models.ts]: Addressed in commit `7cfce28`.

Added regression tests confirming token-profile defaults preserve `preferredModelId` when the scoped registry list is empty, when the registry is missing, and when `availableModelIds` is empty, preventing fallback to canonical Claude models in those paths.

**Comment by jeremymcs** [comment]
## CI needs attention

Changes were pushed in commit `7cfce28`, but CI/CD checks are still failing:

- **windows-portability**: Check run cancelled ([details](https://github.com/open-gsd/gsd-pi/actions/runs/28173153027/job/83443310594))
- **node22-smoke**: Check run cancelled ([details](https://github.com/open-gsd/gsd-pi/actions/runs/28173153027/job/83443310056))
- **ci-gate**: Check run failure ([details](https://github.com/open-gsd/gsd-pi/actions/runs/28173153027/job/83443309767))
- **build**: Check run cancelled ([details](https://github.com/open-gsd/gsd-pi/actions/runs/28173153027/job/83442561127))

Review the failing checks before merging.

File: src/resources/extensions/gsd/model-router.ts [hunk 3]
```diff
@@ -586,14 +609,18 @@ function findModelForTier(
...
-  const eligible = getEligibleModels(tier, availableModelIds, routingConfig);
+  const eligible = getEligibleModels(tier, availableModelIds, routingConfig, preferredModelId);
   if (eligible.length === 0) return undefined;
 
   if (crossProvider) {
     return eligible[0];
   }
 
+  const preferred = findPreferredModelForTier(tier, availableModelIds, preferredModelId);
+  if (preferred && eligible.includes(preferred)) return preferred;
```
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/gsd/model-router.ts]: **Accepted** — queued for a code change.
**In progress** — applying the accepted fix.
**Verifying** — checking the applied changes.
**Resolved** — addressed in commit `7fd1dd5`.

**Comment by jeremymcs** [comment]
Started an automated PR update.

Accepted review feedback is being applied.
Progress and final results will be posted in the related review thread(s).

File: src/resources/extensions/gsd/model-router.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (jeremymcs) [reply src/resources/extensions/gsd/model-router.ts]: Addressed in commit `7fd1dd5`.

Already addressed on the current branch: `cross_provider:false` now filters to Claude same-provider models before honoring the session preferred model, so non-Anthropic preferred models cannot bypass the restriction. Verified with the focused `model-router.test.ts` suite, including the regression case for this review comment.
